### ISSUE TITLE
How to train my own model in image classification

### ISSUE THREAD

**Issue by martasp** [dotnet/machinelearning-samples#200]
I cant find sample of image classification how to train model. I don't want to use pre-trained model from this example https://github.com/dotnet/machinelearning-samples/tree/master/samples/csharp/getting-started/DeepLearning_ImageClassification_TensorFlow

**Comment by CESARDELATORRE** [comment]
@martasp - As built-in ML.NET when creating custom-models, ML.NET supports regular machine learning tasks/algorithms mostly using data, such as multiclass-classification, regression, clustering, etc.

Image classification is not regular machine learning tasks but Deep Learning. 

For Deep Learning, ML.NET provides integration to TensorFlow, but as of today, we just provide the possibility of scoring an existing/trained TensorFlow model.

The simplest way to do image classification is to use Azure Cognitive Services CUSTOM VISION, train an image classification model, export it as a TensorFlow model (.pb file) then you can run/score it with ML.NET:

Here's a sample app I created where I originally created the TensorFlow model from Azure Cognitive Services and exported the model as .pb file:

https://github.com/CESARDELATORRE/TensorFlowImageClassificationWebAPI

Here's how I trained the image-classification in Azure Cognitive Services - Custom Vision:

2.	TensorFlow Model created/trained with Cognitive Services Custom Vision: 

a.	Created 2 labels, one per object/furniture-type trained in the model:

                     i.      001-Green-Meeting-Chair-Redmond

                     ii.      002-High-Metal-Chair-Redmond

b.	Per object, I trained with almost 92 pictures for one chair and around 43 pictures for the other chair, as shown here at thehttps://customvision.ai/ portal:

![image](https://user-images.githubusercontent.com/1712635/51066154-79eba780-15bd-11e9-94f8-8fed1c0e3c13.png)

The .pb model file and labels.txt file (just for testing) are available in the GitHub Repo where I have the .NET code. No need for you to recreate another model in CUSTOM VISION until you want to train your own images.. 

In the repo with the sample code there's Web API project here: 
https://github.com/CESARDELATORRE/TensorFlowImageClassificationWebAPI/tree/master/TensorFlowImageClassificationWebAPI
 
You can directly submit any image for object classification by using PostMan:

![image](https://user-images.githubusercontent.com/1712635/51066260-0a29ec80-15be-11e9-8f06-51728f0e36e0.png)

When you send any image from PostMan, you can see you get the image in the Web API:

![image](https://user-images.githubusercontent.com/1712635/51066288-288fe800-15be-11e9-9e24-3ef722f0a023.png)

After that point, you can use ML.NET code to use the TensorFlow model to score/predict/classify the image.

----

The other choice for creating a custom image classification model would be to use Python+KERAS+TensorFlow, like here (basically using Python-KERAS-TF under the covers):

https://github.com/dotnet-architecture/eShopOnContainersAI/wiki/08.-Computer-Vision:-Scenario-3:-TensorFlow-Custom-Model-setup-and-training

----

In the future we might implement integration to TensorFlow to allow CREATE/TRAINING a TensorFlow model from ML.NET API. But currently that is not possible.


Hope it helps.

**Comment by davidlfox** [comment]
What was your performance like on this project? I trained a model with ~30 images of cats and ~30 images of toasters, and the Precision, Recall, and AP are all 100%.

**Comment by prathyusha12345** [comment]
@martasp I am closing the issue hoping its resolved. 

Closing the issue. Please reopen if you need further details.Thanks

**Comment by ArrowRaider** [comment]
I don't want to use Azure.  I wrote my own TensorFlow model for something unrelated to images.  How can I get it into ML.NET?

**Comment by gvashishtha** [comment]
Try exporting to ONNX and using the ApplyOnnxModel method: https://docs.microsoft.com/en-us/dotnet/api/microsoft.ml.onnxcatalog.applyonnxmodel?view=ml-dotnet
